NOISSUE - Handle authorization errors other than unauthorize

dborovcanin opened the pull request.

dborovcanin committed:
Update CoAP adapter not to return Unauthorize in case of unreachable Clients service

Fix typo in api.go.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Update CoAP adapter README.md

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Add handling errors during authorization

During authorization, possible errors are unauthorized, service unreachable and invalid argument.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Remove usage of lable from code

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Fix logging in case of authorization failure

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

anovakovic01 commented on ws/api/transport.go:
```diff
@@ -55,6 +57,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
+				}
+				return
+			}
 			logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
```
Log error before extracting error code.

anovakovic01 commented on ws/api/transport.go:
```diff
@@ -55,6 +57,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
+			if ok {
+				switch e.Code() {
+				case codes.PermissionDenied:
+					w.WriteHeader(http.StatusUnauthorized)
```
Use forbidden instead.

anovakovic01 commented on ws/api/transport.go:
```diff
@@ -55,6 +57,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
+				case codes.PermissionDenied:
+					w.WriteHeader(http.StatusUnauthorized)
+				case codes.InvalidArgument:
+					w.WriteHeader(http.StatusBadRequest)
```
This case is handled in `authorize` function. Request won't be sent if there is no token. If you want, you can update error there.

anovakovic01 commented on ws/api/transport.go:
```diff
@@ -55,6 +57,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
+				case codes.InvalidArgument:
+					w.WriteHeader(http.StatusBadRequest)
+				default:
+					w.WriteHeader(http.StatusServiceUnavailable)
```
Should client receive service unavailable in default case?

anovakovic01 commented on http/api/transport.go:
```diff
@@ -111,6 +111,18 @@ func encodeResponse(_ context.Context, w http.ResponseWriter, response interface
+	if ok {
+		switch e.Code() {
+		case codes.PermissionDenied:
+			w.WriteHeader(http.StatusUnauthorized)
```
Use forbidden instead.

anovakovic01 commented on http/api/transport.go:
```diff
@@ -111,6 +111,18 @@ func encodeResponse(_ context.Context, w http.ResponseWriter, response interface
+		case codes.PermissionDenied:
+			w.WriteHeader(http.StatusUnauthorized)
+		case codes.InvalidArgument:
+			w.WriteHeader(http.StatusBadRequest)
```
This case is handled in `authorize` function.

anovakovic01 commented on ws/api/transport.go:
```diff
@@ -55,7 +57,19 @@ func handshake(svc ws.Service) http.HandlerFunc {
+				logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
+				switch e.Code() {
+				case codes.PermissionDenied:
+					w.WriteHeader(http.StatusUnauthorized)
```
Use `forbidden` instead.

anovakovic01 commented on ws/api/transport.go:
```diff
@@ -55,7 +57,19 @@ func handshake(svc ws.Service) http.HandlerFunc {
-			logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
+			e, ok := status.FromError(err)
+			if ok {
+				logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
```
Log every error, not just those with code.

dborovcanin committed:
Add forbidden access error handle

Add forbidden status hadling to all services.
Fix tests to correspond newest changes.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

mijicd commented on clients/api/grpc/endpoint_test.go:
```diff
@@ -70,16 +68,19 @@ func TestCanAccess(t *testing.T) {
 		chanID    string
 		id        string
-		err       error
+		respCode  codes.Code
```
Please rename to `code`.

mijicd commented on clients/mocks/security.go:
```diff
@@ -24,7 +28,7 @@ func (hm *hasherMock) Compare(plain, hashed string) error {
 
 func (idp *identityProviderMock) TemporaryKey(id string) (string, error) {
-	if id == "" {
+	if id == "" || strings.Contains(id, "wrong") {
```
What's the purpose of this?

mijicd commented on http/api/transport.go:
```diff
@@ -111,15 +111,28 @@ func encodeResponse(_ context.Context, w http.ResponseWriter, response interface
 }
 
 func encodeError(_ context.Context, err error, w http.ResponseWriter) {
+	e, ok := status.FromError(err)
```
Infline with if and move to the `default` case of the switch below.

mijicd commented on ws/api/transport.go:
```diff
@@ -56,6 +58,18 @@ func handshake(svc ws.Service) http.HandlerFunc {
 		}
 		if err != nil {
 			logger.Warn(fmt.Sprintf("Failed to authorize: %s", err))
+			e, ok := status.FromError(err)
```
Same here.

dborovcanin committed:
Move handling of CanAccess errors to `default` case

Remove `magic` value from client mocks.

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>

dborovcanin committed:
Add InternalServerError handle

Signed-off-by: Dušan Borovčanin <dusan.borovcanin@mainflux.com>
